src/OrleansRuntime/Scheduler/ShedulerExtensions.cs

```diff
@@ -7,64 +7,23 @@ internal static class SchedulerExtensions
     {
         internal static Task<T> QueueTask<T>(this OrleansTaskScheduler scheduler, Func<Task<T>> taskFunc, ISchedulingContext targetContext)
         {
-            var resolver = new TaskCompletionSource<T>();
-            Func<Task> asyncFunc =
-                async () =>
-                {
-                    try
-                    {
-                        T result = await taskFunc();
-                        resolver.TrySetResult(result);
-                    }
-                    catch (Exception exc)
-                    {
-                        resolver.TrySetException(exc);
-                    }
-                };
-
-            // it appears that it's not important that we fire-and-forget asyncFunc() because we wait on the
-            scheduler.QueueWorkItem(new ClosureWorkItem(() => asyncFunc().Ignore()), targetContext);
-            return resolver.Task;
+            var workItem = new AsyncClosureWorkItem<T>(taskFunc);
+            scheduler.QueueWorkItem(workItem, targetContext);
+            return workItem.Task;
         }
 
         internal static Task QueueTask(this OrleansTaskScheduler scheduler, Func<Task> taskFunc, ISchedulingContext targetContext)
         {
-            var resolver = new TaskCompletionSource<bool>();
-            Func<Task> asyncFunc =
-                async () =>
-                {
-                    try
-                    {
-                        await taskFunc();
-                        resolver.TrySetResult(true);
-                    }
-                    catch (Exception exc)
-                    {
-                        resolver.TrySetException(exc);
-                    }
-                };
-            scheduler.QueueWorkItem(new ClosureWorkItem(() => asyncFunc().Ignore()), targetContext);
-            return resolver.Task;
+            var workItem = new AsyncClosureWorkItem(taskFunc);
+            scheduler.QueueWorkItem(workItem, targetContext);
+            return workItem.Task;
         }
 
         internal static Task QueueNamedTask(this OrleansTaskScheduler scheduler, Func<Task> taskFunc, ISchedulingContext targetContext, string activityName = null)
         {
-            var resolver = new TaskCompletionSource<bool>();
-            Func<Task> asyncFunc =
-                async () =>
-                {
-                    try
-                    {
-                        await taskFunc();
-                        resolver.TrySetResult(true);
-                    }
-                    catch (Exception exc)
-                    {
-                        resolver.TrySetException(exc);
-                    }
-                };
-            scheduler.QueueWorkItem(new ClosureWorkItem(() => asyncFunc().Ignore(), () => activityName), targetContext);
```

Comment:
> I was talking both about ALL usages if closure work item

So am I - there are 3 execution paths which pass *any* value for the name, two are constants and one is not. For the one which is not constant, the only usage of *that* code path sets it to a field value - `GrainTimer.Name` and the `GrainTimer` must outlive the work item since the work item holds a reference to it.

> Why do u think on completed request should be scheduled?

I just expected both paths (success or failure) to proceed in the same way. Instead, one is executed inline and the other is executed via the task scheduler. Looking closer I can see that it's correct & required to keep the activation's message pump running.